File: balboa-common/src/main/java/com/socrata/balboa/metrics/data/impl/TimeService.java
package com.socrata.balboa.metrics.data.impl;

/**
 * Returns the current time. Does not go backwards.
 */
public class TimeService {
    long lastTime = 0;

    /**
     * Returns the current time, or the last time returned if
     * the system clock is going backwards. Yes. This happened.
     * Yes. Cassandra apparently requires the timestamp to move
     * forwards, and I need a time-mock anyways... so...
     */
    public long currentTimeMillis() {
        long currentTime = System.currentTimeMillis();
        if (currentTime < lastTime) {
            return lastTime;
        } else {
            lastTime = currentTime;
            return currentTime;
        }

    }
}

File: balboa-common/src/main/scala/com/socrata/balboa/metrics/data/impl/TimeService.scala
package com.socrata.balboa.metrics.data.impl

/**
  * Returns the current time and guarantees that successive calls
  * are well ordered. After a given time has been returned, no
  * lesser time will ever be returned.
  *
  * This is necessary to avoid backwards movement that can
  * sometimes occur in the system clock. Cassandra requires
  * timestamps to move forwards.
 */
class TimeService {
  var lastTime = 0

  def currentTimeMillis(): Long = {
    this.synchronized {
      val currentTime = System.currentTimeMillis()
      if (currentTime < lastTime) {
        lastTime
      } else {
        currentTime
      }
    }
  }
}
